src/main/java/de/tum/in/www1/artemis/service/hestia/structural/StructuralTestCaseService.java

```diff
@@ -0,0 +1,407 @@
+     *
+     * @param programmingExercise The programming exercise
+     * @return The new structural solution entries
+     * @throws StructuralSolutionEntryGenerationException If there was an error while generating the solution entries
+     */
+    public List<ProgrammingExerciseSolutionEntry> generateStructuralSolutionEntries(ProgrammingExercise programmingExercise) throws StructuralSolutionEntryGenerationException {
+        log.debug("Generating the structural SolutionEntries for the following programmingExercise: {} {}", programmingExercise.getId(), programmingExercise.getProjectName());
+
+        var testCases = testCaseRepository.findByExerciseIdWithSolutionEntriesAndActive(programmingExercise.getId(), true);
+        testCases.removeIf(testCase -> testCase.getType() != ProgrammingExerciseTestCaseType.STRUCTURAL);
+
+        // No test cases = no solution entries needed
+        if (testCases.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        // Checkout the solution and test repositories
+        Repository solutionRepository;
+        Repository testRepository;
+        try {
+            var solutionParticipation = solutionProgrammingExerciseParticipationRepository.findByProgrammingExerciseId(programmingExercise.getId());
+            if (solutionParticipation.isEmpty()) {
+                return Collections.emptyList();
+            }
+            solutionRepository = gitService.getOrCheckoutRepository(solutionParticipation.get().getVcsRepositoryUrl(), true);
+            testRepository = gitService.getOrCheckoutRepository(programmingExercise.getVcsTestRepositoryUrl(), true);
+        }
+        catch (InterruptedException | GitAPIException e) {
+            var error = "Error while checking out repositories";
+            log.error(error, e);
+            throw new StructuralSolutionEntryGenerationException(error, e);
+        }
+
+        var classElements = readStructureOracleFile(testRepository.getLocalPath());
+        var solutionClasses = getClassesFromFiles(retrieveJavaSourceFiles(solutionRepository.getLocalPath()));
+
+        // Create new solution entries
+        var newSolutionEntries = Arrays.stream(classElements).flatMap(classElement -> {
+            var packageName = classElement.getStructuralClass().getPackageName();
+            var name = classElement.getStructuralClass().getName();
+            var solutionClass = solutionClasses.get(packageName + "." + name);
+            String filePath;
+            if (solutionClass != null) {
+                filePath = solutionClass.getSource().getURL().getPath().replace(solutionRepository.getLocalPath().toAbsolutePath() + "/", "");
+            }
+            else {
+                filePath = "src/" + packageName.replaceAll("\\.", "/") + "/" + name + ".java";
+            }
+
+            String classSolutionCode = generateCodeForClass(classElement.getStructuralClass(), classElement.getEnumValues(), solutionClass);
```

Comment:
As a first iteration the generation of the code in here is okay, I think. In future, though, defining a common interface for the classes `hestia.structural.Structural*.java` and then implementing something like a visitor would maybe be the more OOP-way of generating the code for multiple elements and then combing it. That would also remove it from this service that at the moment consists of mostly source code generation.

Reply:
I agree.
We will do this in a follow-up PR